lib/manticore/client/trust_strategies.rb

```diff
@@ -0,0 +1,101 @@
+      #   @param coercible [Proc<(Array<OpenSSL::X509::Certificate>,String)>:Boolean]
+      #     A proc that accepts two arguments and returns a boolean value.
+      #     Expects a proc that operates on a pair of parameters:
+      #       @param cert_chain [Enumerable<OpenSSL::X509::Certificate>]: the peer's certificate chain
+      #       @param auth_type [String]: the authentication type based on the client certificate
+      #       @return [Boolean]: true if the certificate can be trusted without verification by the trust manager,
+      #                        false otherwise.
+      #   @example: CA Trusted Fingerprint
+      #      ca_trusted_fingerprint = lambda do |cert_chain, type|
+      #        cert_chain.lazy
+      #                  .map(&:to_der)
+      #                  .map(&::Digest::SHA256.method(:hexdigest))
+      #                  .include?("324a87eebb19265ffb675dc345eb0f3b5d9de3f015159227a00fe552291d4cc4")
+      #      end
+      #      TrustStrategiesInterface.coerce(ca_trusted_fingerprint)
+      def self.coerce(coercible)
+        case coercible
+        when TrustStrategy, nil then coercible
+        when ::Proc             then CustomTrustStrategy.new(coercible)
+        else fail(ArgumentError, "No implicit conversion of #{coercible} to #{self}")
+        end
+      end
+
+      # Combines two possibly-nil TrustStrategiesInterface-coercible objects into a
+      # single TrustStrategy, or to nil if both are nil.
+      #
+      # @param lhs [nil|TrustStrategiesInterface]
+      # @param rhs [nil|TrustStrategiesInterface]
+      # @return [nil,TrustStrategiesInterface]
+      def self.combine(lhs, rhs)
+        return coerce(rhs) if lhs.nil?
+        return coerce(lhs) if rhs.nil?
+
+        CombinedTrustStrategy.new(lhs,rhs)
+      end
+    end
+
+    ##
+    # @api private
+    # A CombinedTrustStrategy can be used to bypass the Trust Manager if
+    # *EITHER* TrustStrategy trusts the provided certificate chain.
+    class CombinedTrustStrategy
+      include TrustStrategiesInterface
+
+      ##
+      # @api private
+      # @see TrustStrategiesInterface::combine
+      def initialize(lhs, rhs)
+        @lhs = lhs
+        @rhs = rhs
+        super()
+      end
+
+      def trusted?(chain, type)
+        @lhs.trusted?(chain, type) || @rhs.trusted?(chain, type)
+      end
+    end
+
+
+    ##
+    # @api private
+    # A CustomTrustStrategy is an org.apache.http.conn.ssl.TrustStrategy
+    # defined with a proc that uses Ruby OpenSSL::X509::Certificates
+    # @see TrustStrategiesInterface::coerce(Proc)
+    class CustomTrustStrategy
+      include TrustStrategiesInterface
+
+      ##
+      # @see TrustStrategiesInterface.coerce(Proc)
+      def initialize(proc)
+        fail(ArgumentError, "2-arity proc required") unless proc.arity == 2
+        @trust_strategy = proc
+      end
+
+      CONVERT_JAVA_CERTIFICATE_TO_RUBY = -> (java_cert) { ::OpenSSL::X509::Certificate.new(java_cert.encoded) }
```

Comment:
Is there a better way to perform this conversion? I note that the JOSSL implementation has a `X509Cert::wrap(ThreadContext context, Certificate cert)` that in [0.12.2](https://github.com/jruby/jruby-openssl/blob/v0.12.2/src/main/java/org/jruby/ext/openssl/X509Cert.java#L153-L156) still calls through `X509Certificate#getEncoded` but in [future releases](https://github.com/jruby/jruby-openssl/blob/master/src/main/java/org/jruby/ext/openssl/X509Cert.java#L158-L166) will use the underlying cert directly.

Reply:
unfortunately there isn't - the `public static IRubyObject wrap` is still primarily for Java callers not Ruby ones.
JOSSL added `Certificate#to_java` conversion lately (will be in 0.13.0) so I've been thinking about the under way around ... there's no code at the moment but there's 2 options:
- have `Certificate.new` accept Java certificates (and `byte[]`)
- add `Certificate.from_java` which would support Java certificates (no `byte[]` for now)
- a `byte[]` convention could be used in numerous places where `String` is accepted so this needs more :cloud: 

second option seems the preferred one atm but I haven't really had time to think about it some more, but it kind of matches what JRuby does e.g. `String#initialize` does not accept `byte[]` there's a `from_java_bytes` method.

which brings us back to this line of code which I do not think would work, let's do:
`(java_cert) { ::OpenSSL::X509::Certificate.new String.from_java_bytes(java_cert.encoded) }` for now

Reply:
The JRuby proxying seems to handle the above correctly, so I'd prefer to leave it as-is and avoid the explicit string creation.

Reply:
interesting, this indeed works, [as](https://github.com/jruby/jruby-openssl/blob/v0.11.0/src/main/java/org/jruby/ext/openssl/StringHelper.java#L111=) a [fallback](https://github.com/jruby/jruby-openssl/blob/v0.11.0/src/main/java/org/jruby/ext/openssl/StringHelper.java#L126=) for the `byte[]` Java proxy implementing `asString()`:

`jruby -e 'bytes = "!42?"; proxy = bytes.to_java(org.jruby.runtime.builtin.IRubyObject); p proxy.asString'`

but there's [no clear intent](https://github.com/jruby/jruby-openssl/blob/v0.11.0/src/main/java/org/jruby/ext/openssl/OpenSSL.java#L337-L341=) (thus the confusion on my part) for the `byte[]` handling when doing `Certificate.new` nor is this tested in JOSSL atm (going to make sure to add tests to not break this functionality).

... seems to work all the way back in 9.1 / 9.2 / 9.3 

am okay with leaving it as is but it's a bit scary that we would use an un-documented / un-tested feature of JOSSL :zipper_mouth_face: